### ISSUE TITLE
Question editor does not work if model joins the same table more than once

### ISSUE THREAD

**Issue by BGough-PIFSC** [metabase/metabase#33427]
### Describe the bug

Question editor does not work with models when the model joins the same table more than 1 time.
 

### To Reproduce

1. Create a SQL based model similar to the following (the query must join to the same table twice):
```
select pr.purchase_report_id, a1.name as created_by, a2.name as updated_by
from purchase_report pr
    join account a1 on a1.account_id = pr.created_by
    join account a2 on a2.account_id = pr.updated_by;
```
2. Adding or removing model metadata has no impact on bug.
3. Create a question that uses the above model (i.e. New -> Question).
4. Click preview button and note that the question works when returning all columns as shown in following screenshot.
![image](https://github.com/metabase/metabase/assets/14341880/3f435963-1630-4bfd-9d5c-47c358660dda)
5. Use the drop down next to the model name to select specific columns.  You will see the updated_by column is listed 2 times as shown in the following screenshot. 
![image](https://github.com/metabase/metabase/assets/14341880/23e0ca13-43d4-4608-9317-2a55b7817acb)

### Expected behavior

Column select drop down should show created_by and updated_by

### Information about your Metabase installation

```JSON
Chrome 115
Windows 10
MySQL
Metabase version 1.47.0 
Metabase hosting environment: Docker
Metabase internal database: MySQL
```

### Severity

Blocking your usage of Model feature entirely

### Additional context

Our shop has a number of databases where you must join the same table multiple times in a single query.  
Our current workaround is to create database views in MySQL instead of using Metabase models.
I also have trouble joining the same table multiple times when using the non-SQL based question editor but this seems to be fixed in version 47.0.

**Comment by ranquild** [comment]
@BGough-PIFSC Could you reproduce this with the Sample Database? I've tried the following query but it worked correctly:
```
select
  ors.USER_ID,
  p1.NAME as created_by,
  p2.NAME as updated_by
from
  ORDERS ors
  join PEOPLE p1 on p1.ID = ors.USER_ID
  join PEOPLE p2 on p2.ID = ors.USER_ID;
```

**Comment by matthelm** [comment]
I tried reproducing this with v0.47.0 and Postgres, but couldn't replicate the issue.
|Creating the model|Querying the model|
|-|-|
|<img width="1017" alt="image" src="https://github.com/metabase/metabase/assets/819476/f5d1d6b1-5297-4b7b-b374-80f0ab06a387">|<img width="769" alt="image" src="https://github.com/metabase/metabase/assets/819476/bf3aea5b-9986-4589-a5d0-ba5da5ab6bef">|

**Comment by sirius-czl** [comment]
Same issue when different columns map to the same table field, it makes the model's schema wrong.
Resolved by remove mapping for one of those columns.

**Comment by ranquild** [comment]
@BGough-PIFSC I reproduced it using your steps, many thanks! 

For backend engineers - we use `Lib.fieldableColumns(query, stageIndex)` + `Lib.displayInfo(query, stageIndex, column)` here and these functions got confused. 

<img width="1728" alt="Screenshot 2023-11-08 at 11 13 53" src="https://github.com/metabase/metabase/assets/8542534/95034dde-bc6b-4bf7-b0bd-e48b249b0bfd">

**Comment by snoe** [comment]
This appears to be a problem with the legacy metadata provider - it is throwing away field information on `"tables"."card__74"` by indexing `fields` by id, but in this case it is insufficient since both fields are mapped to a the product.title field id.

![Image](https://github.com/metabase/metabase/assets/105012/161f29d5-385e-4618-b001-3a6b81f1cecb)

![Image](https://github.com/metabase/metabase/assets/105012/8de2eb9a-f56a-479b-8994-e0a0650bcf49)

**Comment by lbrdnk** [comment]
Following steps in [this comment](https://github.com/metabase/metabase/issues/33427#issuecomment-1690861463) issue is reproducible on master [d419e9a](https://github.com/metabase/metabase/commit/d419e9acc4109c59878f4b6fc51a86e23f7dc051).

**Comment by kamilmielnik** [comment]
Fixed by #44721
